File: dmn-core/src/test/resources/tck/cl3/expected/dmn/0014-loan-comparison/type/TLoanProduct.java
package type;

import java.util.*;

@javax.annotation.Generated(value = {"itemDefinitionInterface.ftl", "tLoanProduct"})
@com.fasterxml.jackson.annotation.JsonPropertyOrder(alphabetic = true)
public interface TLoanProduct extends com.gs.dmn.runtime.DMNType {
    static TLoanProduct toTLoanProduct(Object other) {
        if (other == null) {
            return null;
        } else if (TLoanProduct.class.isAssignableFrom(other.getClass())) {
            return (TLoanProduct)other;
        } else if (other instanceof com.gs.dmn.runtime.Context) {
            TLoanProductImpl result_ = new TLoanProductImpl();
            result_.setLenderName((String)((com.gs.dmn.runtime.Context)other).get("lenderName"));
            result_.setRate((java.math.BigDecimal)((com.gs.dmn.runtime.Context)other).get("rate"));
            result_.setPoints((java.math.BigDecimal)((com.gs.dmn.runtime.Context)other).get("points"));
            result_.setFee((java.math.BigDecimal)((com.gs.dmn.runtime.Context)other).get("fee"));
            return result_;
        } else if (other instanceof com.gs.dmn.runtime.DMNType) {
            return toTLoanProduct(((com.gs.dmn.runtime.DMNType)other).toContext());
        } else {
            throw new com.gs.dmn.runtime.DMNRuntimeException(String.format("Cannot convert '%s' to '%s'", other.getClass().getSimpleName(), TLoanProduct.class.getSimpleName()));
        }
    }

    @com.fasterxml.jackson.annotation.JsonGetter("lenderName")
    String getLenderName();

    @com.fasterxml.jackson.annotation.JsonGetter("rate")
    java.math.BigDecimal getRate();

    @com.fasterxml.jackson.annotation.JsonGetter("points")
    java.math.BigDecimal getPoints();

    @com.fasterxml.jackson.annotation.JsonGetter("fee")
    java.math.BigDecimal getFee();

    default com.gs.dmn.runtime.Context toContext() {
        com.gs.dmn.runtime.Context context = new com.gs.dmn.runtime.Context();
        context.put("lenderName", getLenderName());
        context.put("rate", getRate());
        context.put("points", getPoints());
        context.put("fee", getFee());
        return context;
    }

    default boolean equalTo(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        TLoanProduct other = (TLoanProduct) o;
        if (this.getFee() != null ? !this.getFee().equals(other.getFee()) : other.getFee() != null) return false;
        if (this.getLenderName() != null ? !this.getLenderName().equals(other.getLenderName()) : other.getLenderName() != null) return false;
        if (this.getPoints() != null ? !this.getPoints().equals(other.getPoints()) : other.getPoints() != null) return false;
        if (this.getRate() != null ? !this.getRate().equals(other.getRate()) : other.getRate() != null) return false;

        return true;
    }

    default int hash() {
        int result = 0;
        result = 31 * result + (this.getFee() != null ? this.getFee().hashCode() : 0);
        result = 31 * result + (this.getLenderName() != null ? this.getLenderName().hashCode() : 0);
        result = 31 * result + (this.getPoints() != null ? this.getPoints().hashCode() : 0);
        result = 31 * result + (this.getRate() != null ? this.getRate().hashCode() : 0);
        return result;
    }

    default String asString() {
        StringBuilder result_ = new StringBuilder("{");
        result_.append("fee=" + getFee());
        result_.append(", lenderName=" + getLenderName());
        result_.append(", points=" + getPoints());
        result_.append(", rate=" + getRate());
        result_.append("}");
        return result_.toString();
    }
}

File: dmn-test-cases/standard/tck/1.1/cl3/0014-loan-comparison/translator/expected/kotlin/dmn/type/TLoanProduct.kt
package type

import java.util.*

@javax.annotation.Generated(value = ["itemDefinitionInterface.ftl", "tLoanProduct"])
@com.fasterxml.jackson.annotation.JsonPropertyOrder(alphabetic = true)
@com.fasterxml.jackson.databind.annotation.JsonDeserialize(`as` = type.TLoanProductImpl::class)
interface TLoanProduct : com.gs.dmn.runtime.DMNType {
    @get:com.fasterxml.jackson.annotation.JsonGetter("lenderName")
    val lenderName: String?

    @get:com.fasterxml.jackson.annotation.JsonGetter("rate")
    val rate: kotlin.Number?

    @get:com.fasterxml.jackson.annotation.JsonGetter("points")
    val points: kotlin.Number?

    @get:com.fasterxml.jackson.annotation.JsonGetter("fee")
    val fee: kotlin.Number?

    override fun toContext(): com.gs.dmn.runtime.Context {
        val context = com.gs.dmn.runtime.Context()
        context.add("lenderName", this.lenderName)
        context.add("rate", this.rate)
        context.add("points", this.points)
        context.add("fee", this.fee)
        return context
    }

    fun equalTo(o: Any?): Boolean {
        if (this === o) return true
        if (javaClass != o?.javaClass) return false

        val other = o as TLoanProduct
        if (if (this.fee != null) this.fee != other.fee else other.fee != null) return false
        if (if (this.lenderName != null) this.lenderName != other.lenderName else other.lenderName != null) return false
        if (if (this.points != null) this.points != other.points else other.points != null) return false
        if (if (this.rate != null) this.rate != other.rate else other.rate != null) return false

        return true
    }

    fun hash(): Int {
        var result = 0
        result = 31 * result + (if (this.fee != null) this.fee.hashCode() else 0)
        result = 31 * result + (if (this.lenderName != null) this.lenderName.hashCode() else 0)
        result = 31 * result + (if (this.points != null) this.points.hashCode() else 0)
        result = 31 * result + (if (this.rate != null) this.rate.hashCode() else 0)
        return result
    }

    fun asString(): String {
        val result_ = StringBuilder("{")
        result_.append("fee=" + fee)
        result_.append(", lenderName=" + lenderName)
        result_.append(", points=" + points)
        result_.append(", rate=" + rate)
        result_.append("}")
        return result_.toString()
    }

    companion object {
        @JvmStatic
        fun toTLoanProduct(other: Any?): TLoanProduct? {
            if (other == null) {
                return null
            } else if (other is TLoanProduct?) {
                return other
            } else if (other is com.gs.dmn.runtime.Context) {
                var result_ = TLoanProductImpl()
                var lenderName = other.get("lenderName")
                result_.lenderName = lenderName as String?
                var rate = other.get("rate")
                result_.rate = rate as kotlin.Number?
                var points = other.get("points")
                result_.points = points as kotlin.Number?
                var fee = other.get("fee")
                result_.fee = fee as kotlin.Number?
                return result_
            } else if (other is com.gs.dmn.runtime.DMNType) {
                return toTLoanProduct(other.toContext())
            } else {
                throw com.gs.dmn.runtime.DMNRuntimeException(String.format("Cannot convert '%s' to '%s'", other.javaClass.getSimpleName(), TLoanProduct::class.java.getSimpleName()))
            }
        }
    }
}
